dashboard/models.py

```diff
@@ -172,17 +172,22 @@ def get_data_last_updated(self) -> Optional[datetime]:
         """ Returns the datetime of the last cron run of all courses
             This checks for any courses where the data_last_updated value is null.
 
+            If there are new courses it will return the lower value of the last run date or the earliest start date.
+
         :return: datetime. Either the earliest or None
-        :rtype: None
+        :rtype: None, datetime
         """
         new_courses = self.filter(data_last_updated__isnull=True)
+        existing_courses = self.filter(data_last_updated__isnull=False)
 
+        existing_earliest = existing_courses.earliest("data_last_updated").data_last_updated
         # If there are new courses (courses with no last run) return the earliest time of all
+        new_earliest = None
         if len(new_courses) > 0:
-            return new_courses.earliest_start_datetime()
-        # Otherwise return the latest cron run
-        else:
-            return self.all().latest("data_last_updated").data_last_updated
+            new_earliest = new_courses.earliest_start_datetime()
+
+        # Return the lower value of existing_earliest and new_earliest, otherwise return None
+        return min((val for val in [existing_earliest, new_earliest] if val is not None), default=None)
```

Comment:
Clever way of getting the minimum of the values when they may be `None`.  I've tried to think of a simpler way to do this, but this seems to be the best.

Could `existing_earliest` ever be `None`?  I guess it probably could if the data had never been updated.

Reply:
Thanks, yeah originally I had a bunch of if statements (as Sam's suggested) and was like there has to be something better and it's possible a value slips thought the cracks.

I almost had it but when the List was None it gave another error, luckily [Python 3.4 added a default](`earliest_start_date_of_new_courses` would be None if there are no new courses.) to return if it's None, otherwise it has to be in a try/except. So this seemed perfect. 

Yeah, both or either have a possibility of being None.

Reply:
Yeah, looking back at your conversation with Sam, I see the evolution your code went through.  I avoided looking at the earlier comments when I was giving my review so I could look at it without preconceived opinions.

Reply:
FWIW, the only improvements I could think of for the logic are:

## Remove `is not None`
```suggestion
        return min((v for v in [existing_earliest, new_earliest] if v), default=None)
```
This is slightly less easier to read, though.

## Use `filter()`
```suggestion
        return min(filter(None, (existing_earliest, new_earliest)), default=None)
```
I think `filter()` is pretty self-explanatory.

Reply:
I guess with the default "is not none" isn't needed. 

I read about using [filter verses a for loop here. ](https://stackoverflow.com/a/53212104/3708872)

It sounds biggest difference is a simple filter like this will filter out zeroes whereas the for loop doesn't. You'd need a more complex function or lambda. In this case I don't **think** we'd have to worry about this that problem since it's doing it based on datetime. Otherwise they probably have the same result and mostly based on syntax preference. It doesn't matter much to me either way.

Reply:
Seems fine going with filter to me, I'm okay with whatever "Commit suggestion" is being pressed here.

Reply:
I committed the filter suggestion.